### PR TITLE
Expand `compiletime.ops`

### PR DESCRIPTION
**Pull request by soronpo**
This PR expands `compiletime.ops` as follows:
* Adds `ops.string.{Length, Substring, Matches}`
* Adds `ops.int.NumberOfLeadingZeros`
* Adds `ops.long` as equivalent to `ops.int`, but for `Long` operations
* Adds `ops.float` and `ops.double`
* Adds `ops.any.IsConst`
* Deprecates `ops.int.ToString` (from 3.2.x) and moves it to `ops.any.ToString`
* Fixes constant termrefs types not being recognized (e.g.: `one.type + two.type`, where `one` and `two` are final vals).

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala
Comment (MaximeKjaer) [inline compiler/src/dotty/tools/dotc/core/Types.scala:4181]: I suppose this is the bit that fixes constant termrefs types not being recognized?

File: compiler/src/dotty/tools/dotc/core/Types.scala
Comment (MaximeKjaer) [inline compiler/src/dotty/tools/dotc/core/Types.scala:4220]: For consistency with the other methods, I think we should:

- Make `isConst` take a parameter `tp: Type` instead of automatically applying to `args.head`
- (nitpick: skip a line around the function)

File: library/src/scala/compiletime/ops/int.scala
Comment (MaximeKjaer) [inline library/src/scala/compiletime/ops/int.scala:184]: In favor of this.

File: tests/neg/singleton-ops-int.scala
Comment (MaximeKjaer) [inline tests/neg/singleton-ops-int.scala:14]: This is nice!

File: compiler/src/dotty/tools/dotc/core/Types.scala
Comment (MaximeKjaer) [inline compiler/src/dotty/tools/dotc/core/Types.scala:4244]: I think I'd be in favor of using `args(0)`, `args(1)` and `args(2)` instead of `.head`, `(1)` and `.last`. I'd want `constantFold3` to throw an exception if it is applied when there are < 3 args. We of course already have protections in place, both with type-checking and the `if nArgs == x` checks, but I think it doesn't hurt to be a tad more defensive.

(same thing in `constantFold2AB`)

File: library/src/scala/compiletime/ops/string.scala
Comment (MaximeKjaer) [inline library/src/scala/compiletime/ops/string.scala:40]: This is quite cool actually! Quite a powerful feature to have in the type system.

File: compiler/src/dotty/tools/dotc/core/Types.scala
Comment (MaximeKjaer) [inline compiler/src/dotty/tools/dotc/core/Types.scala:4367]: Could we encounter a `IndexOutOfBoundsException`s here? It's probably better to throw a `TypeError` in that case.

File: library/src/scala/compiletime/ops/string.scala
Comment (MaximeKjaer) [inline library/src/scala/compiletime/ops/string.scala:31]: Should we also add the `Substring` without an `IEnd`? Would overloading the type alias work?

**Review by MaximeKjaer — COMMENTED** [review]
In general, this looks good to me! But seeing that I'm not currently very active in the day-to-day Dotty development, I'm adding @OlivierBlanvillain as a reviewer, so that we have the approval of one of the core maintainers too.

I do also have a higher-level question about Strings in particular: should we aim to have full support of all the methods in `java.lang.String`? We're missing methods like `charAt`, `startsWith`, etc. Some of these methods also take objects, such as `format` which takes a `String` or a `Locale` for the `locale` parameter. We can't support `Locale`, but we could support `String`.

File: library/src/scala/compiletime/ops/string.scala
Comment (soronpo) [reply library/src/scala/compiletime/ops/string.scala:31]: No, overloading will not work, so using the one with `IEnd` is the best we have. Those who need the short version can create an alias with `IEnd = Length[S]-1`.

**Comment by soronpo** [comment]
> I do also have a higher-level question about Strings in particular: should we aim to have full support of all the methods in `java.lang.String`? We're missing methods like `charAt`, `startsWith`, etc. Some of these methods also take objects, such as `format` which takes a `String` or a `Locale` for the `locale` parameter. We can't support `Locale`, but we could support `String`.

I currently do not support `Char`. Seems redundant to me, and can always be converted to an operation on a string in length of 1. So `CharAt` is redundant if we have `Substring`. Additionally, `Matches` covers more use cases, and users can create their own `StartsWith` etc. using type alias compositions. I think this PR covers most common use cases, and if we implement the SIP https://contributors.scala-lang.org/t/pre-sip-custom-dependent-operation-types/5236 , then the rest can be supported rather easily.

File: compiler/src/dotty/tools/dotc/core/Types.scala
Comment (soronpo) [reply compiler/src/dotty/tools/dotc/core/Types.scala:4181]: Yes. I'll add a comment there.

File: compiler/src/dotty/tools/dotc/core/Types.scala
Comment (soronpo) [reply compiler/src/dotty/tools/dotc/core/Types.scala:4367]: Yes, I'll change it so that all `constantFoldXX` will trap exceptions from executing their evaluation function and change it to a type error.

**Comment by soronpo** [comment]
@smarter since we are currently in RC1, can we still get this in 3.1.0 if there will be RC2?

**Comment by smarter** [comment]
No, only fixes to regressions are backported into the release branch.

**Comment by smarter** [comment]
Though the added operations could be added in 3.1.1 as `@experimental` (but ops.int.ToString cannot be deprecated until there's a non-experimental replacement).

**Comment by soronpo** [comment]
OK, then the deprecation for `int.ToString` needs to change to `3.1.1`. I'll handle that as well.

**Comment by mbovel** [comment]
I am very interested by this PR, I was going to add the same fix for getting the underlying types of termrefs 😃

What is the status? Could we merge this soon or are there any blockers?

**Comment by smarter** [comment]
It needs to be reviewed and the added definitions need to be marked `@experimental` so we can drop the "needs-minor-release" label.

**Comment by soronpo** [comment]
> 
> 
> I am very interested by this PR, I was going to add the same fix for getting the underlying types of termrefs 😃
> 
> What is the status? Could we merge this soon or are there any blockers?

I'll try to complete this soon.

File: compiler/src/dotty/tools/dotc/core/Types.scala
Comment (johnynek) [reply compiler/src/dotty/tools/dotc/core/Types.scala:4426]: BTW: Can you share a scastie of how a macro could accomplish what I am trying to do here (type-level records that can interop with existing scala and java code, e.g. spark)? To me, the type-level sorting seems to be exactly what I want.

Comment (soronpo) [reply compiler/src/dotty/tools/dotc/core/Types.scala:4426]: The PR is not a place for this discussion. You can open a new discussion on this repo's discussion section and I'll try to reply when I find the time.

**Review by OlivierBlanvillain — APPROVED** [review]
LGTM!

**Comment by OlivierBlanvillain** [comment]
> BTW: I think this PR really brings up a weakness of match types: we shouldn't need to have a PR in the compiler to implement these functions.

@johnynek IMO it's perfectly normal to require compiler support for type-level operations on primitive types. In the long term, the goal should be to be on par with term-level operations on primitives. If somebody is interested in working on a follow-up PR to add support for `Char` and other `String` operation, it's very likely to get in!

File: tests/neg/singleton-ops-int.scala
Comment (soronpo) [reply tests/neg/singleton-ops-int.scala:74]: > What's wrong with that test?

Nothing was wrong. At first I wanted to remove all `int.ToString` tests because it was deprecated, but then I decided not to remove the tests until it was removed completely. So perhaps I forgot to bring these test back as well. I could change the commit, but I don't think it's a big deal.

File: library/src/scala/compiletime/ops/long.scala
Comment (soronpo) [reply library/src/scala/compiletime/ops/long.scala:8]: It's just missing a colon at the end. I'll fix.

File: library/src/scala/compiletime/ops/int.scala
Comment (smarter) [inline library/src/scala/compiletime/ops/int.scala:186]: @soronpo I think this warning will be shown even before 3.2.0 is released which means users will have to choose between a deprecated and experimental, could you comment out the deprecated annotation for now?

Comment (soronpo) [reply library/src/scala/compiletime/ops/int.scala:186]: @smarter why would it show before? The version is specified. Is there a bug in `@deprecated`? If so, we should fix this instead.

**Pull request merged**
